### PR TITLE
Backend Search

### LINKED ISSUE #1112: Support Backend Search
**Is your feature request related to a problem? Please describe.**
Support Full Backend Search using the existing tags based API:
- [x] Extend existing api to support a start/end parameter and for queriers to respect that.
- [x] Add a config option to split traces between ingesters and backend.
- [x] Documentation
- [x] Add support for Google Cloud Functions

### PR DESCRIPTION
**Pull request by joe-elliott**
**What this PR does**:
Adds support for full backend search to the `/api/search` endpoint. If parameters `start` and `end` are provided Tempo will perform a full backend search by scanning the blocks in parallel. This PR also includes code to run and configuration to use a Google Cloud Function to perform the scanning of blocks.

Changes:
- Additional query-frontend params related to search. `default_result_limit` and `max_result_limit` were moved from the querier config to here:
```
query_frontend:
  search:
    concurrent_jobs: <int>
    target_bytes_per_job: <int>
    default_result_limit: <int>
    max_result_limit: <int>
    max_duration: <duration>
    query_ingesters_within_min: <duration>
    query_ingesters_within_max: <duration>
```
- Additional storage parameters related to search:
```
storage:
  trace:
    search:
      chunk_size_bytes: <int>
      prefetch_trace_count: <int>
```
- Full Backend and Serverless integration tests
- Fixes on some flakey integration tests by using `cortex_e2e.WaitMissingMetrics`
- Testing on model.Matches()

**Which issue(s) this PR fixes**:
Fixes #1112 

Even though we are clearing #1112 as part of this PR backend search is far from complete. A new issue will be created "Backend search production readiness" (#1175) that will track future search work.

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: modules/frontend/config.go [hunk 1]
```diff
@@ -21,6 +23,15 @@ func (cfg *Config) RegisterFlagsAndApplyDefaults(prefix string, f *flag.FlagSet)
 	cfg.MaxRetries = 2
 	cfg.QueryShards = 20
 	cfg.TolerateFailedBlocks = 0
+	cfg.Search = SearchSharderConfig{
+		QueryIngestersWithinMin: 15 * time.Minute,
+		QueryIngestersWithinMax: time.Hour,
+		DefaultLimit:            20,
+		MaxLimit:                0,
+		MaxDuration:             61 * time.Minute,
```
Comment (mdisibio) [inline modules/frontend/config.go:31]: **question:** Why 61 minutes and not `time.Hour` ?

File: modules/frontend/searchsharding.go [hunk 2]
```diff
@@ -69,11 +71,17 @@ func (r *searchResponse) addResponse(res *tempopb.SearchResponse) {
 	r.mtx.Lock()
 	defer r.mtx.Unlock()
 
-	r.results.Traces = append(r.results.Traces, res.Traces...)
+	for _, t := range res.Traces {
+		// just take the first
+		if _, ok := r.resultsMap[t.TraceID]; !ok {
+			r.resultsMap[t.TraceID] = t
```
Comment (mdisibio) [inline modules/frontend/searchsharding.go:77]: **suggestion**: This probably needs to dedupe/combine search results, for example the start/end/rootspan may be spread across multiple pages/blocks, and the individual segments that matched don't have the complete picture.  But not critical, can be done later (if we decide it is needed).    Add todo for now?

File: modules/frontend/searchsharding.go [hunk 3]
```diff
@@ -289,18 +342,90 @@ func (s *searchSharder) shardedRequests(ctx context.Context, metas []*backend.Bl
...
 			subR.Header.Set(user.OrgIDHeaderName, tenantID)
 
-			q := subR.URL.Query()
-			q.Add(api.URLParamBlockID, blockID)
-			q.Add(api.URLParamStartPage, strconv.Itoa(startPage))
-			q.Add(api.URLParamTotalPages, strconv.Itoa(pagesPerQuery))
+			subR, err := api.BuildSearchBlockRequest(subR, &tempopb.SearchBlockRequest{
+				BlockID:       blockID,
+				StartPage:     uint32(startPage),
+				TotalPages:    uint32(pagesPerQuery),
```
Comment (mdisibio) [inline modules/frontend/searchsharding.go:348]: **nit** The mismatch between these names is a little confusing.  `TotalPages` seems like the total number of pages in the block, but it is actually the number of pages for the query to inspect.  Might think about switching to Start/End page numbers, which conceptually lines up well with start/end times, or renaming to something like `PageCount`

File: modules/frontend/searchsharding.go [hunk 4]
```diff
@@ -289,18 +342,90 @@ func (s *searchSharder) shardedRequests(ctx context.Context, metas []*backend.Bl
...
 
 	return reqs, nil
 }
+
+// queryIngesterWithin returns a new start and end time range for the backend as well as an http request
+// that covers the ingesters. If nil is returned for the http.Request then there is no ingesters query.
+// if the returned start == the returned end then no further querying is necessary
+func (s *searchSharder) ingesterRequest(ctx context.Context, tenantID string, parent *http.Request, searchReq *tempopb.SearchRequest) (uint32, uint32, *http.Request, error) {
+	now := time.Now()
+	ingesterMin := uint32(now.Add(-s.cfg.QueryIngestersWithinMin).Unix())
```
Comment (mdisibio) [inline modules/frontend/searchsharding.go:373]: **suggestion**: The logic here is a bit hard to follow.  The intent as described is:
```
now -> min: ingester search only
min -> max: ingesters and backend
max -> past: backend only
```

In that case there are 2 thoughts:
* It seems `QueryIngestersWithinMin` is not really ingester-related at all, as the ingesters are queried no matter the value.  It controls whether the backend is searched.   I think it might be clearer to have the two params like:
  * `QueryIngestersUntil: 1h`
  * `QueryBackendAfter: 15m`
* Refactor `ingesterRequest` to only involve ingester, and move the backend start/end trimming elsewhere (i.e. new method)

**Review by kvrhdn — COMMENTED** [review]
Really nice work consolidating all those search functions!

File: docs/tempo/website/configuration/_index.md [hunk 5]
```diff
@@ -196,6 +196,45 @@ query_frontend:
...
+        # min and max will be queried from both the ingesters and the backend.
+        # (default: 15m)
+        [query_ingesters_within_min: <duration>]
+
+        # This duration controls the max cutoff between relying on ingesters and the backend for recent trace information.
+        # If it is set to 1h then data within the last 1h will be queried from the ingesters and all previous data will
+        # be queried from the backend. Notice there is overlap with query_ingesters_within_min. The period of time between
+        # min and max will be queried from both the ingesters and the backend.
+        # (default: 1h)
+        [query_ingesters_within_max: <duration>]
```
[the remark this answers is not in the document]
Comment (joe-elliott) [reply docs/tempo/website/configuration/_index.md]: Agree to adding a timeline of all relevant config options, but its outside the scope of this PR. I'm not even sure I could name them all off the top of my head.

File: modules/frontend/searchsharding.go (hunk 3 above)
Comment (joe-elliott) [reply modules/frontend/searchsharding.go]: Renamed to `pagesToSearch`

**Review by mdisibio — APPROVED** [review]
Excellent work! LGTM
